File: lab6/C++/src/SupplyRequestDetail.cpp
#include <string>
#include <algorithm>
#include "SupplyRequestDetail.hpp"

SupplyRequestDetail::SupplyRequestDetail(const SupplyRequestDetail& other) : 
    m_id(rand()), m_materialType(other.m_materialType), m_count(other.m_count),
    m_caliber(other.m_caliber), m_fuelType(other.m_fuelType), m_vehicleType(other.m_vehicleType){
}

SupplyRequestDetail::SupplyRequestDetail(int id, MaterialType type, float count)
    : m_id(id), m_materialType(type), m_count(count) {
}

SupplyRequestDetail& SupplyRequestDetail::WithCaliber(Caliber caliber) {
    m_caliber = caliber;
    return *this;
}

SupplyRequestDetail& SupplyRequestDetail::WithFuelType(FuelType fuelType) {
    m_fuelType = fuelType;
    return *this;
}

SupplyRequestDetail& SupplyRequestDetail::WithVehicleType(VehicleType vehicleType) {
    m_vehicleType = vehicleType;
    return *this;
}

int SupplyRequestDetail::GetId() const {
    return m_id;
}

MaterialType SupplyRequestDetail::GetSupplyType() const {
    return m_materialType;
}

Caliber SupplyRequestDetail::GetCaliber() const {
    return m_caliber;
}

VehicleType SupplyRequestDetail::GetVehicleType() const {
    return m_vehicleType;
}

FuelType SupplyRequestDetail::GetFuelType() const {
    return m_fuelType;
}

float SupplyRequestDetail::GetCount() const {
    return m_count;
}

void SupplyRequestDetail::SetCount(float count) {
    m_count = count;
}

SupplyRequestDetail& SupplyRequestDetail::operator +=(float count) {
    m_count += count;
    return *this;
}

SupplyRequestDetail& SupplyRequestDetail::operator -=(float count) {
    m_count -= std::min(m_count, count);
    return *this;
}

std::string SupplyRequestDetail::ToString() const {

    std::string result = MaterialTypeToString(m_materialType) + ": ";

    switch (m_materialType) {
    case MaterialType::eAmmunition:
        result += CaliberToString(m_caliber) + ", " + std::to_string(m_count) + " шт";
        break;
    case MaterialType::eFuel:
        result += FuelTypeToString(m_fuelType) + ", " + std::to_string(m_count) + " л";
        break;
    case MaterialType::eVehicle:
        result += VehicleTypeToString(m_vehicleType) + ", " +
            FuelTypeToString(m_fuelType) + ", " +
            std::to_string(m_count) + " шт";
        break;
    case MaterialType::eWeapon:
        result += CaliberToString(m_caliber) + ", " + std::to_string(m_count) + " шт";
        break;
    }

    return result;
}

std::string SupplyRequestDetail::MaterialTypeToString(MaterialType mtype) {
    return m_materialTypes.at(mtype);
}
std::string SupplyRequestDetail::CaliberToString(Caliber caliber) {
    return m_calibers.at(caliber);
}
std::string SupplyRequestDetail::FuelTypeToString(FuelType ftype) {
    return m_fuelTypes.at(ftype);
}

std::string SupplyRequestDetail::VehicleTypeToString(VehicleType vtype) {
    return m_vehicleTypes.at(vtype);
}


const std::map<MaterialType, std::string> SupplyRequestDetail::m_materialTypes {
    {MaterialType::eAmmunition, "Боеприпасы"},
    {MaterialType::eFuel, "Топливо"},
    {MaterialType::eVehicle, "Транспорт"},
    {MaterialType::eWeapon, "Вооружение"}
};

// Инициализация словаря калибров
const std::map<Caliber, std::string> SupplyRequestDetail::m_calibers = {
    {Caliber::e545mm, "5.45мм"},
    {Caliber::e122mm, "122мм"}
};

// Инициализация словаря типов топлива
const std::map<FuelType, std::string> SupplyRequestDetail::m_fuelTypes {
    {FuelType::eGasoline, "Бензин"},
    {FuelType::eDiesel, "Дизель"}
};

// Инициализация словаря типов транспорта
const std::map<VehicleType, std::string> SupplyRequestDetail::m_vehicleTypes = {
    {VehicleType::eTank, "Танк"},
    {VehicleType::eArmoredVehicle, "Бронемашина"},
    {VehicleType::eMotorbike, "Мотоцикл"},
    {VehicleType::eSelfPropelledLauncher, "Самоходная установка"}
};

File: lab6/Java/src/request/SupplyRequestDetail.java
package request;

import enums.*;
import java.util.ArrayList;
import java.util.Random;

public class SupplyRequestDetail implements Cloneable{
    int id;
    SupplyType supplyType;
    Caliber caliber;
    VehicleType vehicleType;
    FuelType fuelType;
    float count;
    
    public SupplyRequestDetail(int id, SupplyType supplyType, float count){
        
        if(count < 1){
            throw new IllegalArgumentException("Недопустимое количество");
        }        

        this.id = id;
        this.supplyType = supplyType;
        setCount(count);
    }

    public SupplyRequestDetail withCaliber(Caliber caliber){
        this.caliber = caliber;
        return this;
    }
    public SupplyRequestDetail withFuelType(FuelType fuelType){
        this.fuelType = fuelType;
        return this;
    }

    public SupplyRequestDetail withVehicleType(VehicleType vehicleType){
        this.vehicleType = vehicleType;
        return this;
    }

    public int getId(){
        return id;
    }

    public SupplyType getSupplyType(){
        return supplyType;
    }

    public Caliber getCaliber(){
        return caliber;
    }

    public VehicleType getVehicleType(){
        return vehicleType;
    }

    public FuelType getFuelType(){
        return fuelType;
    }

    public float getCount(){
        return count;
    }

    public final void setCount(float count){
        this.count = switch(supplyType){
            case AMMUNITION, VEHICLE, WEAPON-> Math.round(count);
            default->count;
        };
    }

    @Override
    public String toString(){
        String result = SupplyTypeToString(supplyType) + ": ";

        result += switch (supplyType) {
            case AMMUNITION, WEAPON ->  CaliberToString(caliber) + ", " + count + " шт";
            case FUEL -> FuelTypeToString(fuelType) + ", " + count + " л";
            case VEHICLE -> VehicleTypeToString(vehicleType) + ", " + FuelTypeToString(fuelType) + ", " + count + "шт";
        };

        return result;
    }
    @Override
    public SupplyRequestDetail clone() throws CloneNotSupportedException{
        var clone = (SupplyRequestDetail)super.clone();
        clone.id = new Random().nextInt();
        return clone;
    }

    public static String toString(ArrayList<SupplyRequestDetail> requestDetails){
        String result = "";
        for (SupplyRequestDetail detail: requestDetails) {
            result += detail.toString().concat("\n");
        }

        return result;
    }

    static String SupplyTypeToString(SupplyType stype){
        return switch(stype){
            case AMMUNITION -> "Боеприпасы";
            case FUEL -> "Топливо";
            case VEHICLE -> "Транспорт";
            case WEAPON -> "Вооружение";
            default -> "Неизвестно";
        };        
    }

    static String CaliberToString(Caliber caliber){
        return switch(caliber){
            case C_122MM -> "122мм";
            case C_545MM -> "5.45мм";
            default -> "Неизвестно";
        }; 
    }

    static String FuelTypeToString(FuelType ftype){
        return switch(ftype){
            case GASOLINE -> "Бензин";
            case DIESEL -> "Дизель";
            default -> "Неизвестно";
        };        
    }

    static String VehicleTypeToString(VehicleType vtype){
        return switch(vtype){
            case ARMORED_VEHICLE -> "Бронемашина";
            case TANK -> "Танк";
            case MOTORBIKE -> "Мотоцикл";
            default -> "Неизвестно";
        };
    }

}
